File: worker/src/contracts/telemetry.ts
export type TelemetryEvent = {
  schemaVersion: "telemetry.v1";
  eventId: string;
  deviceId: string;
  sensorId: string;
  sequence: number;
  measuredAt: string;
  source?: string;
  payload: {
    kind: "temperature" | "light" | "distance" | "button";
    value: number | boolean;
    unit: string;
  };
  correlationId?: string;
  sessionId?: string;
};

export function parseTelemetry(buffer: Buffer): TelemetryEvent {
  const value = JSON.parse(buffer.toString("utf8")) as unknown;
  if (!isTelemetryEvent(value)) {
    throw new Error("Invalid telemetry event");
  }
  return value;
}

function isTelemetryEvent(value: unknown): value is TelemetryEvent {
  if (!value || typeof value !== "object") return false;
  const event = value as Partial<TelemetryEvent>;
  if (event.schemaVersion !== "telemetry.v1") return false;
  if (!event.eventId || !event.deviceId || !event.sensorId) return false;
  if (typeof event.sequence !== "number" || !Number.isInteger(event.sequence) || event.sequence < 0) return false;
  if (!event.measuredAt || Number.isNaN(Date.parse(event.measuredAt))) return false;
  if (!event.payload || typeof event.payload !== "object") return false;
  if (!["temperature", "light", "distance", "button"].includes(event.payload.kind)) return false;
  if (typeof event.payload.value !== "number" && typeof event.payload.value !== "boolean") return false;
  return typeof event.payload.unit === "string" && event.payload.unit.length > 0;
}

File: api/internal/kafka/telemetry.go
package kafka

import (
	"errors"
	"time"
)

type TelemetryPayload struct {
	Kind  string      `json:"kind"`
	Value any         `json:"value"`
	Unit  string      `json:"unit"`
}

type TelemetryEvent struct {
	SchemaVersion string           `json:"schemaVersion"`
	EventID       string           `json:"eventId"`
	DeviceID      string           `json:"deviceId"`
	SensorID      string           `json:"sensorId"`
	Sequence      int64            `json:"sequence"`
	MeasuredAt    string           `json:"measuredAt"`
	Source        string           `json:"source,omitempty"`
	Payload       TelemetryPayload `json:"payload"`
	CorrelationID string           `json:"correlationId,omitempty"`
	SessionID     string           `json:"sessionId,omitempty"`
}

func (event TelemetryEvent) Validate() error {
	if event.SchemaVersion != "telemetry.v1" {
		return errors.New("schemaVersion must be telemetry.v1")
	}
	if event.EventID == "" || event.DeviceID == "" || event.SensorID == "" {
		return errors.New("eventId, deviceId, and sensorId are required")
	}
	if event.Sequence < 0 {
		return errors.New("sequence must be non-negative")
	}
	if _, err := time.Parse(time.RFC3339, event.MeasuredAt); err != nil {
		return errors.New("measuredAt must be an RFC3339 timestamp")
	}
	if event.Payload.Kind == "" || event.Payload.Unit == "" {
		return errors.New("payload.kind and payload.unit are required")
	}
	return nil
}
